R/db-helpers.R

```diff
@@ -157,12 +157,21 @@ get_src_tbl_names <- function(src, schema = NULL, dbname = NULL) {
     names_table <- get_names_table_mariadb(con)
   }
 
+  # Use smart default for `.names`, if it wasn't provided
+  names_pattern <- if (length(schema) == 1) {
+    names %||% "{.table}"
+  } else {
+    names %||% "{.schema}.{.table}"
+  }
```

Comment:
Will this be surprising for users of this function?

Reply:
You know this package's audience far better than I do! So I'll ultimately take your advice here. But I'll share why I believe this default isn't _too_ surprising for users with either a very little bit of SQL/database experience, or some slightly broader tidyverse experience:

#### With (a very little) SQL experience:

* In SQL (and assuming we are working within a single database), entities are usually specified as `schema.table`, e.g. `select * from schema.table`
* Although it is strongly discouraged when writing SQL directly, it's possible (for at least some versions of SQL) to omit the schema name - in which case, the table is searched for in the user's "default-schema" (or, in each schema in turn along their "search path" in some versions e.g. Postgres)
* The default-schema has a default value (e.g. `dbo` for SQL Server, `public` for Postgres), but it is possible for a user to override it via an explicit command (e.g. StackOverflow instructions [here](https://dba.stackexchange.com/questions/42955/query-without-having-to-specify-tables-schema) for SQL Server)
* So with this in mind, I think users fall into one of three categories:
  - "Huh, what's a schema?": no problem, they use the default default-schema for whichever database they connect to (i.e. they do not specify `schema =` in `dm_from_con()`), and can reach all the tables in that schema via a single-part name like `table`
  - "I only need things in one particular schema": they either intentionally use the default schema, or specify a single non-default schema in `dm_from_con()` - but either way, they are conscious that they are using a single schema of their choice, i.e. they have knowingly chosen their personal default-schema and can therefore use single-part names like `table`
  - "I need multiple schemas!": they specify multiple schemas in `dm_from_con()` - they are conscious of this decision, and are effectively setting their "search path" for tables. The default naming pattern is guaranteed to give unique two-part names like `schema.table`, but the user _can_, against best practice, choose to use a different pattern, e.g. `{.table}` to generate one-part names. Only the first table on the search path with each name will be kept, with a warning raised about tables not kept, thanks to #1789.
  
#### With some broader tidyverse experience:

There are already well-established examples of using glue patterns and "special variables" (for us, `.schema` and `.table`) for construction of object names elsewhere in the tidyverse - for example:

* `.names` in `dplyr::across()` - which has a near-identical "smart default" to ours here, i.e. switches from one-part to two-part names in order to avoid ambiguity when multiple "sources" are involved (schemas in our case, functions in `across()`'s case)
* `names_glue` in `tidyr::pivot_wider()`